versioned_hdf5/slicetools.pyx

```diff
@@ -234,6 +238,125 @@ cpdef build_data_dict(dcpl, raw_data_name: str):
 
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.infer_types(True)
+cpdef tuple[np.ndarray, np.ndarray] build_slab_indices_and_offsets(
+    dcpl,
+    shape: tuple[int, ...],
+    chunk_size: tuple[int, ...],
+):
+    """Build the slab_indices and slab_offsets arrays for a StagedChangesArray that
+    accesses a versioned virtual dataset.
+
+    :param dcpl: the dataset creation property list of the versioned dataset
+    :return: tuple of (slab_indices, slab_offsets) arrays, each with the same shape as
+    the chunks in the virtual dataset.
+    """
+    cdef hsize_t[NPY_MAXDIMS] chunk_size_arr
+    cdef hsize_t[NPY_MAXDIMS] nchunks_arr
+    cdef hsize_t[NPY_MAXDIMS] h5_start
+    cdef hsize_t[NPY_MAXDIMS] scratch  # Dumping ground for unused outputs
+
+    nchunks = []
+    ndim = len(shape)
+    for j in range(ndim):
+        chunk_size_j: hsize_t = chunk_size[j]
+        assert chunk_size_j > 0
+        chunk_size_arr[j] = chunk_size[j]
+        nchunks_j = ceil_a_over_b(shape[j], chunk_size_j)
+        nchunks_arr[j] = nchunks_j
+        nchunks.append(nchunks_j)
+
+    # Default for missing chunks is to point to slabs[0],
+    # which is a single chunk covered in fill_value
+    cdef np.ndarray slab_indices = np.zeros(nchunks, dtype=np_hsize_t)
+    cdef np.ndarray slab_offsets = np.zeros(nchunks, dtype=np_hsize_t)
+    indices_strides = slab_indices.strides
+    slab_indices_data = slab_indices.data
+    slab_offsets_data = slab_offsets.data
+
+    with phil:
+        # Newly created sparse datasets completely full of fill_value are not virtual.
+        # FIXME Error getting virtual sources on empty virtual dataset
+        # https://github.com/h5py/h5py/issues/1660
+        if dcpl.get_layout() != h5d.VIRTUAL or 0 in shape:
+            return slab_indices, slab_offsets
+
+        dcpl_id: hid_t = dcpl.id
+        virtual_count: size_t = dcpl.get_virtual_count()
+
+        with nogil:
+            # Iterate over all chunks except full ones
+            for chunk_i in range(virtual_count):
+                # Calculate offset of the chunk, IN BYTES, relative to the start of the
+                # flattened buffer of slab_indices and slab_offsets
+                vspace_id = H5Pget_virtual_vspace(dcpl_id, chunk_i)
+                if vspace_id == H5I_INVALID_HID:
+                    raise HDF5Error()
+                try:
+                    vspace_type = H5Sget_select_type(vspace_id)
+                    if vspace_type != H5S_sel_type.H5S_SEL_HYPERSLABS:
+                        raise ValueError(f"Unexpected {vspace_type=}")
+
+                    # Get chunk index in the virtual dataset
+                    if H5Sget_regular_hyperslab(
+                        vspace_id,
+                        h5_start,
+                        scratch,  # IGNORE - stride (always 1)
+                        scratch,  # IGNORE - count
+                        scratch,  # IGNORE - block (typically, but not always, 1)
```

Comment:
Inconsequential as long as the virtual dataset is contiguous, which is always the case